### PR TITLE
[SPARK-48281][SQL] Alter string search logic for UTF8_BINARY_LCASE collation (StringInStr, SubstringIndex)

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
String searching in UTF8_BINARY_LCASE now works on character-level, rather than on byte-level. For example: `instr("İ", "i")`; now returns 0, because there exists no `start, len` such that `lowercase(substring("İ", start, len)) == "i"`.

### Why are the changes needed?
Fix functions that give unusable results due to one-to-many case mapping when performing string search under UTF8_BINARY_LCASE (see example above).

### Does this PR introduce _any_ user-facing change?
Yes, behaviour of `instr` and `substring_index` expressions is changed for edge cases with one-to-many case mapping.

### How was this patch tested?
New unit tests in `CollationSupportSuite`.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 1]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
       return UTF8String.EMPTY_UTF8;
     }
 
-    UTF8String lowercaseString = string.toLowerCase();
     UTF8String lowercaseDelimiter = delimiter.toLowerCase();
 
     if (count > 0) {
-      int idx = -1;
+      // search left to right (note: the start code point is inclusive)
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:437]:
```suggestion
      // Search left to right (note: the start code point is inclusive).
```
Please use regular sentences as comments.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 2]
```diff
@@ -183,6 +320,22 @@ public static int findInSet(final UTF8String match, final UTF8String set, int co
...
+   * @param pattern the string to be searched for
+   * @param start the start position for searching (in the target string)
+   * @return the position of the first occurrence of pattern in target, if not found, -1 returned.
+   */
+  public static int lowercaseIndexOf(final UTF8String target, final UTF8String pattern,
+      final int start) {
+    if (pattern.numChars() == 0) return 0;
+    return lowercaseFind(target, pattern.toLowerCase(), start);
+  }
+
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:338]: I am not reviewing changes from this point and above. I have reviewed them in https://github.com/apache/spark/pull/46511

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 3]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
...
-          count--;
-        } else {
-          // can not find enough delim
-          return string;
-        }
-      }
-      if (idx == 0) {
-        return UTF8String.EMPTY_UTF8;
+        matchLength = lowercaseFind(string, lowercaseDelimiter, matchLength + 1);
+        if (matchLength > MATCH_NOT_FOUND) count--; // found a delimiter
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:441]:
```suggestion
        if (matchLength != MATCH_NOT_FOUND) --count; // Found a delimiter.
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 4]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
...
+        else return string; // cannot find enough delimiters in the string
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:442]:
```suggestion
        else return string; // Cannot find enough delimiters in the string.
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 5]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
...
       }
-      byte[] bytes = new byte[idx];
-      copyMemory(string.getBaseObject(), string.getBaseOffset(), bytes, BYTE_ARRAY_OFFSET, idx);
-      return UTF8String.fromBytes(bytes);
-
+      if (matchLength == 0) return UTF8String.EMPTY_UTF8;
+      return string.substring(0, matchLength);
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:445]:
```suggestion
      return string.substring(0, matchLength);
```
The `UTF8String.substring` method does what you want for you.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 6]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
...
     } else {
-      int idx = string.numBytes() - delimiter.numBytes() + 1;
+      // search right to left (note: the end code point is exclusive)
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:447]:
```suggestion
      // Search right to left (note: the end code point is exclusive).
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 7]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
...
       while (count > 0) {
-        idx = lowercaseString.rfind(lowercaseDelimiter, idx - 1);
-        if (idx >= 0) {
-          count--;
-        } else {
-          // can not find enough delim
-          return string;
-        }
+        matchLength = lowercaseRFind(string, lowercaseDelimiter, matchLength - 1);
+        if (matchLength > MATCH_NOT_FOUND) count--; // found a delimiter
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:452]:
```suggestion
        if (matchLength != MATCH_NOT_FOUND) --count; // Found a delimiter.
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 8]
```diff
@@ -278,47 +431,29 @@ public static UTF8String lowercaseSubStringIndex(final UTF8String string,
...
-      if (idx + delimiter.numBytes() == string.numBytes()) {
-        return UTF8String.EMPTY_UTF8;
-      }
-      int size = string.numBytes() - delimiter.numBytes() - idx;
-      byte[] bytes = new byte[size];
-      copyMemory(string.getBaseObject(), string.getBaseOffset() + idx + delimiter.numBytes(),
-        bytes, BYTE_ARRAY_OFFSET, size);
-      return UTF8String.fromBytes(bytes);
+      if (matchLength == string.numChars()) return UTF8String.EMPTY_UTF8;
+      return string.substring(matchLength, string.numChars());
```
Comment (mkaravel) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:456]:
```suggestion
      return string.substring(matchLength, string.numChars());
```
Same here.

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 9]
```diff
@@ -567,8 +567,26 @@ public void testStringInstr() throws SparkException {
...
+    assertStringInstr("ai̇oxXİo", "Xx", "UNICODE_CI", 5);
+    assertStringInstr("aİoi̇oxx", "XX", "UNICODE_CI", 7);
+    assertStringInstr("i̇", "i", "UTF8_BINARY_LCASE", 1); // != UNICODE_CI
+    assertStringInstr("i̇", "\u0307", "UTF8_BINARY_LCASE", 2); // != UNICODE_CI
+    assertStringInstr("i̇", "İ", "UTF8_BINARY_LCASE", 1);
+    assertStringInstr("İ", "i", "UTF8_BINARY_LCASE", 0);
+    assertStringInstr("İoi̇o12", "i̇o", "UTF8_BINARY_LCASE", 1);
+    assertStringInstr("i̇oİo12", "İo", "UTF8_BINARY_LCASE", 1);
+    assertStringInstr("abİoi̇o", "i̇o", "UTF8_BINARY_LCASE", 3);
+    assertStringInstr("abi̇oİo", "İo", "UTF8_BINARY_LCASE", 3);
```
Comment (mkaravel) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:587]: Can we also add?
```
assertStringInstr("abI\u0307oi̇o", "İo", "UTF8_BINARY_LCASE", 3);
assertStringInstr("abİoi̇o", "\u0307o", "UTF8_BINARY_LCASE", 6);
```

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 10]
```diff
@@ -798,6 +816,30 @@ public void testSubstringIndex() throws SparkException {
...
+    assertSubstringIndex("ai̇bİoi̇o12i̇oİo", "İo", 3, "UNICODE_CI", "ai̇bİoi̇o12");
+    assertSubstringIndex("ai̇bİoi̇o12i̇oİo", "i̇o", 3, "UNICODE_CI", "ai̇bİoi̇o12");
+    assertSubstringIndex("abi̇12", "i", 1, "UTF8_BINARY_LCASE", "ab"); // != UNICODE_CI
+    assertSubstringIndex("abi̇12", "\u0307", 1, "UTF8_BINARY_LCASE", "abi"); // != UNICODE_CI
+    assertSubstringIndex("abi̇12", "İ", 1, "UTF8_BINARY_LCASE", "ab");
+    assertSubstringIndex("abİ12", "i", 1, "UTF8_BINARY_LCASE", "abİ12");
+    assertSubstringIndex("ai̇bi̇oİo12İoi̇o", "İo", -4, "UTF8_BINARY_LCASE", "İo12İoi̇o");
+    assertSubstringIndex("ai̇bi̇oİo12İoi̇o", "i̇o", -4, "UTF8_BINARY_LCASE", "İo12İoi̇o");
+    assertSubstringIndex("ai̇bİoi̇o12i̇oİo", "İo", -4, "UTF8_BINARY_LCASE", "i̇o12i̇oİo");
+    assertSubstringIndex("ai̇bİoi̇o12i̇oİo", "i̇o", -4, "UTF8_BINARY_LCASE", "i̇o12i̇oİo");
```
Comment (mkaravel) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:838]: Can we also add interesting test cases where there is no match ?

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 11]
```diff
@@ -798,6 +816,30 @@ public void testSubstringIndex() throws SparkException {
...
+    assertSubstringIndex("ai̇bi̇oİo12İoi̇o", "İo", 3, "UTF8_BINARY_LCASE", "ai̇bi̇oİo12");
+    assertSubstringIndex("ai̇bi̇oİo12İoi̇o", "i̇o", 3, "UTF8_BINARY_LCASE", "ai̇bi̇oİo12");
+    assertSubstringIndex("ai̇bİoi̇o12i̇oİo", "İo", 3, "UTF8_BINARY_LCASE", "ai̇bİoi̇o12");
+    assertSubstringIndex("ai̇bİoi̇o12i̇oİo", "i̇o", 3, "UTF8_BINARY_LCASE", "ai̇bİoi̇o12");
```
Comment (mkaravel) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:842]: Same for these.

**Review by mkaravel — APPROVED** [review]
Thank you for addressing my comments.
LGTM!

**Comment by uros-db** [comment]
updated PR description & title
@cloud-fan ready for review

**Review by uros-db — COMMENTED** [review]
updated PR description & title
@cloud-fan ready for review

**Comment by mkaravel** [comment]
@uros-db Please see my comment about the PR description in https://github.com/apache/spark/pull/46511
The same comment applies to the PR description for this PR.

**Review by dbatomic — APPROVED** [review]

**Comment by cloud-fan** [comment]
please fix merge conflicts.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**